src/wired/dataclasses/injector.py

```diff
@@ -0,0 +1,110 @@
+from dataclasses import dataclass, fields, Field, MISSING
+from typing import get_type_hints, Optional, Tuple
+
+from wired import ServiceContainer
+
+from wired.dataclasses.models import Context
+
+
+@dataclass()
+class Injector:
+    """ Introspect dataclass and get arguments from container """
+    container: ServiceContainer
+
+    def __call__(self,
+                 target,
+                 singletons: Optional[Tuple] = (),
+                 context: Optional[Context] = None
+                 ):
+
+        container = self.container
+
+        # Don't need to construct this target from a dataclass,
+        # it's a singleton instance already in the container
+        if target in singletons:
+            instance = container.get(target)
+            return instance
+
+        # Make the args dict that we will construct dataclass with
+        args = {}
+
+        # If not passed in, get the context from the container
+        if context is None:
+            try:
+                context: Context = container.get(Context)
+            except LookupError:
+                # It's ok to have context=None
+                pass
```

Comment:
Can you help me understand the difference between this and `container.context`? It's not quite clicking for me yet.

Reply:
Sometimes, in the process of injector-resolution, you have something asking for a different context than the container. (I doubt that's a satisfactory answer.)

Reply:
Also, `register_dataclass` lets you [register a dataclass for a context](https://github.com/pauleveritt/wired/blob/dataclasses/tests/dataclasses/unit/test_registration.py#L62).

This is then used in the [decorator closure for the generated factory](https://github.com/pauleveritt/wired/blob/dataclasses/src/wired/dataclasses/registration.py#L25).

Reply:
Wired doesn't know anything about pyramid or requests. Each container does have a `context` attribute which is basically like `context = injected(ServiceContainer, attr='context')`. This would also be the instance of the context, versus the type that you pass to `register_dataclass`, but they are I'm pretty sure what you're looking at here.

Reply:
I know wired is distinct from pyramid but I was trying to follow a Pyramid pattern which I thought wired did as well, meaning, there's something called a "context". 

How do I associate a particular context instance with a container? Is it `.bind`?

Reply:
`registry.create_container(context=...)` is the first way. Other than that you can later create other pseudo-containers that are bound to a context via `.bind` which returns a new container, sharing the same caches, but with a different bound context. `container.context` can change per-factory based on how the factory is registered. If this isn't clear I'll have to think of a better way to describe it - it's probably one of the most subtle parts of wired.

Reply:
I think the `context` arg should be removed here, and the `models.Context` idea dropped for this iteration.

Reply:
I removed most of Context but left it in as something that people could ask for during injection.

I could change it so that Context isn't something that people can use in their dataclasses (factories.) It just seems...weird. I'm registering this factory for a context=FrenchCustomer but I can't ask for that context, I have to wire up some other mechanism to hand me the context, when it's right there in `container.context`.

Reply:
Who is registering the Context service? Why isn’t the injector just using container.context? You already made the ServiceContainer injectable so it’s just the context attr on that object. Doesn’t need to be separate really. The wired context is not a service it’s more like a scope.